Compare destination image manifest copy optimization/feature

gjenkins8 opened the pull request:
This PR optimizes image copies for the scenario where the destination image is expected to exist.

The scenario occurs for a common usage of `skopeo sync`: when periodically synchronizing two registries. Where the set of source images are (mostly) unchanged, and already exist at the destination. For this scenario, fetching the destination image's manifest, and checking if it matches the source manifest. And skipping the 'copy' (ie. layer content checks and manifest update) if equivalent provides a performance improvement.

This check should be an opt-in. As the default scenario for image copying is assumedly copying an image to a new destination. And the manifest fetch is pessimistic in this case. The copy is also never skipped if the copy requires signing or encryption (through more work could be done to skip the copy if the signing or encryption are un-changed).

More details / related issue:
https://github.com/containers/skopeo/issues/1021

Thanks!

Sample output when copy is skipped:
```
INFO[0004] Copying image tag 2/42                        from="docker://registry.company.com/subpath/foo:0.0.18" to="docker://registry.company.com/synctest/foo:0.0.18"
Skipping: image already present at destination
```

mtrmac reviewed (commented):
Thanks!

I haven’t checked how this integrates into the full process, for now just quick notes:
- Please sign your commit per https://github.com/containers/image/blob/master/CONTRIBUTING.md#sign-your-prs
- Looking at the test failures, failure of `destImageSource.GetManifest` should not be fatal (it is fine, and quite expected, for the destination not to contain any image before a copy happens).
- The opt-in to do this check should eventually be exposed in `copy.Options`.

TomSweeneyRedHat commented on copy/copy.go:
```diff
@@ -577,6 +577,29 @@ func (c *copier) copyOneImage(ctx context.Context, policyContext *signature.Poli
 		}
 	}
 
+	destExistsCheck := true // TBD: enable configuration
```
Maybe something in [containers.conf](https://github.com/containers/common/blob/master/pkg/config/containers.conf) eventually?

mtrmac replied:
No, this is not a system-wide decision; whether to do the check (= optimistically assume that the copy is a no-op) it is individual to the specific operation (is this a push of a just-built image that does not exist anywhere else vs. a daily sync of a rarely-changing large set of repositories).

So, for c/image this should be only a Go option with no user configuration, and I guess _most_ callers should leave this off. `skopeo sync` can start with hard-coding to on. Eventually we might encounter commands or cases where the right default is not obvious based on just the command, and where the difference is big enough, and that software can expose the option in the CLI or in individual config files.

gjenkins8 replied:
Thanks @TomSweeneyRedHat , @mtrmac, I have created an option `DestinationManifestCheck` in `copy.Options` (suggestions on naming completely welcome, there is likely a more descriptive name out there)

mtrmac replied:
Maybe something like `AssumeDestinationExists`? The API should describe the situation / assumption / desired behavior, not details of the implementation.

(Anyway, I’m afraid I still need to look into the details of correctness.)

TomSweeneyRedHat commented on copy/copy.go:
```diff
@@ -577,6 +577,29 @@ func (c *copier) copyOneImage(ctx context.Context, policyContext *signature.Poli
+
+			retManifest, retManifestType, err = destImageSource.GetManifest(ctx, nil)
+			if err != nil {
+				return nil, "", "", errors.Wrapf(err, "Error reading manifest from dest image %s", transports.ImageName(destImageSource.Reference()))
```
```suggestion
				return nil, "", "", errors.Wrapf(err, "Error reading manifest from destination image %s", transports.ImageName(destImageSource.Reference()))
```

mtrmac replied:
BTW most of the code in this file does not say source/destination at all, the image name + operation should be sufficient.

mtrmac replied:
… but anyway we need to carefully deal with correctness first, and see what text is necessary for that, before fine-tuning it.

(e.g. — note to self — we still need to do the copy if the source contains signatures. There may be more.)

TomSweeneyRedHat commented on copy/copy.go:
```diff
@@ -577,6 +577,29 @@ func (c *copier) copyOneImage(ctx context.Context, policyContext *signature.Poli
+			}
+
+			if bytes.Compare(sourceManifest, retManifest) == 0 {
+				c.Printf("Dest image manifest compares equal\n")
```
Soft suggestion here and above for dest -> destination
```suggestion
				c.Printf("Destination image manifest compares equal\n")
```

gjenkins8 replied:
Thanks! I've also amend my initial PR note with a sample output

TomSweeneyRedHat commented:
Will need a rebase at some point @gjenkins8 , nice start!

gjenkins8 commented:
Thanks @mtrmac @TomSweeneyRedHat for the inital pass. I have tidied up the changes:

> Please sign your commit per https://github.com/containers/image/blob/master/CONTRIBUTING.md#sign-your-prs
✅ Done (apologies)

> Looking at the test failures, failure of destImageSource.GetManifest should not be fatal (it is fine, and quite expected, for the destination not to contain any image before a copy happens).
✅ Done, via only proceeding when non-nil error for both of these calls (`NewImageSource ` & `GetManifest`).

>  The opt-in to do this check should eventually be exposed in `copy.Options`
✅ Done -- `DestinationManifestCheck`

I've also tidied up the verbiage, and posted a sample output to the initial PR note

gjenkins8 commented:
> Anyway, I’m afraid I still need to look into the details of correctness.

Thanks @mtrmac -- I have marked as ready for review for when you have time. Thanks!

gjenkins8 commented:
@mtrmac -- in the tests:

```
Running skopeo --tls-verify=false --policy /tmp/policy.json322663611 copy atomic:localhost:5006/myns/cosigned:cosigned dir:/tmp/signatures-dest332000716
copy_test.go:835:
    assertSkopeoSucceeds(c, "", "--tls-verify=false", "--policy", policy, "copy", "atomic:localhost:5006/myns/cosigned:cosigned", dirDest)
utils.go:62:
    c.Assert(err, check.IsNil, check.Commentf("%s", out))
... value *exec.ExitError = &exec.ExitError{ProcessState:(*os.ProcessState)(0xc0002227a0), Stderr:[]uint8(nil)} ("exit status 1")
time="2020-09-20T04:48:31Z" level=fatal msg="Source image rejected: Invalid GPG signature: gpgme.Signature{Summary:128, Fingerprint:\"8BD82FEFF8CB2EAADA1270B8F2BC068064B5AAF3\", Status:gpgme.Error{err:0x9}, Timestamp:time.Time{wall:0x0, ext:63736174108, loc:(*time.Location)(0x55cc6e3bf680)}, ExpTimestamp:time.Time{wall:0x0, ext:62135596800, loc:(*time.Location)(0x55cc6e3bf680)}, WrongKeyUsage:false, PKATrust:0x0, ChainModel:false, Validity:0, ValidityReason:error(nil), PubkeyAlgo:1, HashAlgo:8}"
```

But I didn't understand why my change would cause the source image to be rejected?

mtrmac commented:
As a very quick guess without actually checking, skipping the copy also skips the opportunity to add a signature via `--sign-by`. I guess some _earlier_ copy was optimized out, causing the image to not be signed as expected; this error complains about a different signature that does exist; if there were a valid signature, that complaint would not be reported to the user.

gjenkins8 commented:
@mtrmac -- I have done the naive thing, and only attempt to skip copying iff signing is not requested (`options.SignBy == ""`).

I don't know enough about signing to know if copying can be skipped (manifests compare equal), but the destination image can be re-signed. I presume it can somehow? But also before trying to implement such, I think it is worth asking how we would like to proceed:

Would we consider skipping when not signing to be good enough for a first cut? Then look to add skipping when signing in later? (presuming possible). Or would you think these should both be delivered together?

Also, my PR is a little awkward. In order to get integration/test coverage, I need to 'force' `options.DestinationManifestCheck`  to be true (as it is essentially acting as a feature flag). But of course, this would be don't by the client (`skopeo`) in practice.

How would you like to proceed? I can remove the 'feature enablement' from this PR, and Skopeo can be updated to set `options.DestinationManifestCheck = true`? Or, I can make a new PR to add `options.DestinationManifestCheck`, a second PR to `skopeo` to enable the option. And this PR would be the 3rd to implement the behaviour?

gjenkins8 commented on copy/copy.go:
```diff
@@ -546,6 +551,25 @@ func (c *copier) copyOneImage(ctx context.Context, policyContext *signature.Poli
 	}
 
+	options.DestinationManifestCheck = true // TODO: Forced enablement for initial PR testing
+	if options.DestinationManifestCheck && options.SignBy == "" {
```
Should verify that `options.DestinationManifestCheck` is not set, if `options.SignBy` is set? Since both of these options are exclusive (with the current implementation).

The alternative would be to leave this decision to the client (which might choose to only set `options.DestinationManifestCheck` if `options.SignBy` is not set). And alway skip the copy exclusively based upon `options.DestinationManifestCheck`.

I don't think `options.DestinationManifestCheck && options.SignBy == "" ` is the right approach, as it allows the client to request skipping the manifest copy when that is not possible.

mtrmac replied:
The way I think about it, the caller is not making a _request_ that the code here must honor — the caller does not want the `copy.Image` call to _fail_ just because a full image must be copied. The caller is only is providing `AssumeImageAlreadyExistsAtDestination` information that this code could completely ignore, or use in any way it wants to optimize the behavior. 

From that point of view, it is perfectly fine for the client to provide “image probably already exists” information, together with a request for an operation (format conversion, encryption, …) which we don’t currently support in the optimization. Maybe such an optimization can be implemented in the future; maybe not. The caller can provide that information just the same, and if some future optimization arrives, the caller will benefit at that time.

gjenkins8 replied:
> The way I think about it, the caller is not making a request that the code here must honor

Thanks, nice. I *was* thinking about it like this. But I agree, this is more of an 'optimization hint'.

mtrmac commented on copy/copy.go:
```diff
@@ -546,6 +551,25 @@ func (c *copier) copyOneImage(ctx context.Context, policyContext *signature.Poli
 	}
 
+	options.DestinationManifestCheck = true // TODO: Forced enablement for initial PR testing
+	if options.DestinationManifestCheck && options.SignBy == "" {
```
The full set of conditions AFAICS:
- Caller opts in
- No need to edit the image/manifest (per the `reflect.DeepEqual` check in `copyUpdatedConfigAndManifest`, which is non-trivial enough that it should probably be split into a separate `ic.noPendingManifestUpdates` helper instead of copy&pasted):
  - No manifest conversion is needed
  - No need to update the embedded Docker reference
- No need to copy existing signatures (`len(sigs) == 0`, or alternatively all of these signatures are known to exist at the destination as well — that can be left to a future PR)
- No new signature (`opts.SignBy == ""`)
- No OCI encryption/decryption (`copier.oci*Config`)

I’m a bit torn on whether to have the conditions inlined here or in a helper function; I'm leaning towards inlining, to make the logic easier to see and to make it more difficult to overlook that this optimization exists.

The “destination already contains the desired manifest” check, OTOH, is conceptually trivial and would be better split into a helper function, to make `copyOneImage` a little less gargantuan.

mtrmac commented on copy/copy.go:
```diff
@@ -546,6 +551,25 @@ func (c *copier) copyOneImage(ctx context.Context, policyContext *signature.Poli
+	if options.DestinationManifestCheck && options.SignBy == "" {
+		destImageSource, err := c.dest.Reference().NewImageSource(ctx, options.DestinationCtx)
+		if err == nil {
+			retManifest, retManifestType, err = destImageSource.GetManifest(ctx, nil)
```
This should specify `targetInstance` so that it benefits multi-arch images as well (and correctly handles updates to a tag between single-arch/multi-arch images).

mtrmac commented on copy/copy.go:
```diff
@@ -546,6 +551,25 @@ func (c *copier) copyOneImage(ctx context.Context, policyContext *signature.Poli
+				}
+
+				if bytes.Equal(sourceManifest, retManifest) {
+					c.Printf("Destination image manifest compares equal\n")
```
Maybe
```suggestion
					c.Printf("Destination already contains the expected manifest, not copying the image again\n")
```
or something a bit succinct — but the goal is that the reader understands why no layer/manifest copies show up in the debug output.

mtrmac replied:
… a bit _more_ succinct, I mean.

gjenkins8 replied:
`"Destination copy skipped as expected image already present"` ?

mtrmac replied:
Probably `s/Destination copy/Copy/`, and maybe even `s| as|,|`; I’m not a native speaker and I’m not sure how much the use of “as” to mean “because” is common.

mtrmac reviewed (commented):
Finally a full review.

mtrmac commented on copy/copy.go:
```diff
@@ -546,6 +551,25 @@ func (c *copier) copyOneImage(ctx context.Context, policyContext *signature.Poli
+		destImageSource, err := c.dest.Reference().NewImageSource(ctx, options.DestinationCtx)
+		if err == nil {
+			retManifest, retManifestType, err = destImageSource.GetManifest(ctx, nil)
+			if err == nil {
```
Non-blocking: For the two error cases above, it would probably be useful to at least log them in the debug output.

gjenkins8 replied:
I've added these, but perhaps I should remove?

They will log if the destination image doesn't exist (albeit at debug level of course), which is a fairly common/expected scenario.

mtrmac commented:
> @mtrmac -- I have done the naive thing, and only attempt to skip copying iff signing is not requested (`options.SignBy == ""`).
> 
> I don't know enough about signing to know if copying can be skipped (manifests compare equal), but the destination image can be re-signed.

Signatures are “attachments” to an image, so it is possible to first create an image, and later add signatures (or even more signatures) to it. So, the presence of a manifest does not automatically imply no need to copy signatures.

For `docker://`, it would be possible to just skip the layer+config+manifest copy, and then add signatures; but right now, the `ImageDestination` interface specifies that layer/config/manifest/signature updates must all happen in that order, so it is not valid to just update signatures. Some cases do arise from time to time where it would be interesting to relax this restriction, but I think we shouldn’t block this PR on an API change like that.

It _would_ be possible, if `options.SignBy` is not set but `sigs` exists, for the optimization to read signatures at the destination (the `ImageSource` interface has no ordering requirements, unlike `ImageDestination`), and to skip the image copy if both the manifest and the sets of signatures are satisfactory. Unless this actually comes up in your use, I think it’s fine to defer that to a separate PR.

> Also, my PR is a little awkward. In order to get integration/test coverage, I need to 'force' `options.DestinationManifestCheck` to be true (as it is essentially acting as a feature flag). But of course, this would be don't by the client (`skopeo`) in practice.

ACK; that’s a great way to test it now, we just shouldn’t merge it like that.

> How would you like to proceed? I can remove the 'feature enablement' from this PR, and Skopeo can be updated to set `options.DestinationManifestCheck = true`? Or, I can make a new PR to add `options.DestinationManifestCheck`, a second PR to `skopeo` to enable the option. And this PR would be the 3rd to implement the behaviour?

The 3-PR version would still have to wait for an update of c/image in Skopeo, so I don’t think it speeds anything up. Let’s
- test this PR with the hard-coded enablement
- Remove the enablement, and merge into c/image
- (Make a new c/image release? That would be cleaner but more overhead.)
- Vendor c/image into Skopeo (either a release, or a git commit)
- In the same vendoring PR, or later, set the flag in `skopeo sync` (and add extra tests?)

The 3-PR version would be helpful if extra Skopeo tests found problems in the implementation; but in the meantime we would have a do-nothing flag in c/image, possibly even in a release. Maybe we can afford to be optimistic and assume that there won’t be huge problems only found later.

gjenkins8 commented on copy/copy.go:
```diff
@@ -182,6 +182,11 @@ type Options struct {
+	// If the destination image manifest exists, and compares equal to the source image manifest, the copy is skipped as the destination already matches the aource.
+	// Otherwise the copy proceeds as normal.
+	// This is a useful optimization to avoid if the destination image is expected to already exist, such as repeated `skopeo sync` between registries.
+	DestinationManifestCheck bool
```
I came up with:
```
	// When OptimizeDestinationImageAlreadyExists is set, optimize the copy assuming that the destination image already
	// exists (and is equivalent). Making the eventual (no-op) copy more performant for this case. Enabling the option
	// is slightly pessimistic if the destination image doesn't exist, or is not equivalent.
	OptimizeDestinationImageAlreadyExists bool
```

gjenkins8 commented on copy/copy.go:
```diff
@@ -633,6 +669,26 @@ func (c *copier) copyOneImage(ctx context.Context, policyContext *signature.Poli
 	// If encrypted and decryption keys provided, we should try to decrypt
 	ic.diffIDsAreNeeded = ic.diffIDsAreNeeded || (isEncrypted(src) && ic.c.ociDecryptConfig != nil) || ic.c.ociEncryptConfig != nil
 
+	options.OptimizeDestinationImageAlreadyExists = true // TODO: PR: Remove forced enablement for PR testing
```
NOTE: Remove before merging

gjenkins8 commented:
@mtrmac -- Thanks for the detailed review! *I think* I managed to understand and address your comments -- from checks to messages. But PTAL to see if I missed anything, or any room for improvement.

Unfortunately, since I reworked basically all of the code. The majority of prior PR comments are now outdated. Not sure if you would want to continue the existing 'threads' or start new ones against the new code (up to you, I don't mind).

gjenkins8 commented on copy/copy.go:
```diff
@@ -633,6 +669,26 @@ func (c *copier) copyOneImage(ctx context.Context, policyContext *signature.Poli
+		}
+	}
+
 	if err := ic.copyLayers(ctx); err != nil {
```
Is it a concern that `ic.copyLayers()` updates `ic. manifestUpdates`, and so can change the result of `ic.noPendingManifestUpdates()`?  But the check to skip the copy on [L678](https://github.com/containers/image/pull/1041/files#diff-5f78baafa6d34438aba66ed64f4b16d5R678) will never see those updates?

I assume it isn't, since if no layers need to be copied because the manifest is equivalent, then `ic.copyLayers()` would not update `ic.manifestUpdates`.

But I also can't say I am a fan of this implicit contract/coupling between these bits of functionality.

mtrmac replied:
Yes, it is “a concern”.

It would of course be nicer if we could rely on the `ic.manifestUpdates` coming out of `ic.copyLayers` — but avoiding `ic.copyLayers` is the whole point, we can’t make some of them without initiating the layer reads via `GetBlob`.

Still, it would probably be worth it to make this a bit harder to miss, maybe by adding a
```go
// WARNING: If you are adding new reasons to change ic.manifestUpdates, update also the OptimizeDestinationImageAlreadyExists short-circuit conditions
```
before the update at the end of `ic.copyLayers`, and a similar
```go
// WARNING: If you are adding new reasons to change the blob, …
```
somewhere in `copyBlobFromStream`, maybe either immediately after the “The copying happens through a pipeline of connected io.Readers.” comment or close to (above?) the “Deal with layer compression/decompression if necessary” one.

gjenkins8 replied:
Above added

gjenkins8 commented:
@mtrmac -- gentle reminder / review appreciated please

mtrmac commented:
@gjenkins8 Thanks for the update. I’m on PTO this week, I’ll return to this afterwards.

ChristianCiach commented:
I've built Skopeo from source to include this PR, because we really need an optimization like this.

Unfortunately, when using your patch, `Destination copy skipped as expected image already present` never gets logged. After adding some debug logging I can see that `ic.noPendingManifestUpdates()` is always true. I think this is already discussed at https://github.com/containers/image/pull/1041/files#r501891650 .

After removing `!ic.noPendingManifestUpdates()` from the `if`-condition, everything works as expected and the sync is hugely accelerated.

gjenkins8 commented:
note: CI is failing due to: https://github.com/containers/image/issues/1107

rhatdan commented:
@gjenkins8 Could you rebase and try again?

gjenkins8 commented:
> @gjenkins8 Could you rebase and try again?

@rhatdan -- looks good now. Thanks!

gjenkins8 commented:
@mtrmac -- if you have some spare time, a review would be muchly appreciated please

Also, reminder of https://github.com/containers/image/pull/1041/files#r500609569 (so it isn't forgotten / is removed before merge)

gjenkins8 commented:
@ChristianCiach -- fyi, https://github.com/containers/image/pull/1041/commits/a35942fcffb1ce747b057b6a8a99beb98d63fcd6 was a fix for a bug we discovered, if you still have a custom build.

ChristianCiach commented:
@gjenkins8 Thanks for the heads up! I will try to build it again some time next week.

mtrmac commented on copy/copy.go:
```diff
@@ -361,6 +365,44 @@ func supportsMultipleImages(dest types.ImageDestination) bool {
 	return false
 }
 
+// compareImageDestinationManifestEqual compares the `src` and `dest` image's manifests, returning the 'true' if they compare equal.
```
`checkIfDestinationManifestIsEqual` ?

“returning true and information about the manifest if…”

mtrmac commented on copy/copy.go:
```diff
@@ -361,6 +365,44 @@ func supportsMultipleImages(dest types.ImageDestination) bool {
 }
 
+// compareImageDestinationManifestEqual compares the `src` and `dest` image's manifests, returning the 'true' if they compare equal.
+// The destination manifest is always fetched.
```
This is not the case on error paths. “This usually fetches the destination manifest”? Or maybe just remove this comment.

mtrmac commented on copy/copy.go:
```diff
@@ -361,6 +365,44 @@ func supportsMultipleImages(dest types.ImageDestination) bool {
+// compareImageDestinationManifestEqual compares the `src` and `dest` image's manifests, returning the 'true' if they compare equal.
+// The destination manifest is always fetched.
+func compareImageDestinationManifestEqual(ctx context.Context, options *Options, src types.Image, dest types.ImageDestination) (bool, []byte, string, digest.Digest, error) {
+
```
(Absolutely non-blocking: Maybe drop this line.)

mtrmac commented on copy/copy.go:
```diff
@@ -361,6 +365,44 @@ func supportsMultipleImages(dest types.ImageDestination) bool {
+		return false, nil, "", "", nil
+	}
+
+	destManifest, destManifestType, err := destImageSource.GetManifest(ctx, nil)
```
Shouldn’t this use `targetInstance`? https://github.com/containers/image/pull/1041#discussion_r499155705

gjenkins8 replied:
oops, thanks! I missed that comment.

I also want to discuss test coverage, but I'll do this in the main PR comment section (so discussion doesn't get lost when this comment becomes outdated )

mtrmac commented on copy/copy.go:
```diff
@@ -361,6 +365,44 @@ func supportsMultipleImages(dest types.ImageDestination) bool {
+		return false, nil, "", "", errors.Wrapf(err, "Error calculating manifest digest")
+	}
+
+	logrus.Debugf("Comparing source and destintion manifest digests: srcManifestDigest=%v, destManifestDigest=%v", srcManifestDigest, destManifestDigest)
```
```suggestion
	logrus.Debugf("Comparing source and destination manifest digests: srcManifestDigest=%v, destManifestDigest=%v", srcManifestDigest, destManifestDigest)
```
or maybe just something like
```suggestion
	logrus.Debugf("Source and destination manifest digests: %v vs. %v", srcManifestDigest, destManifestDigest)
```
to be less verbose in a many-image `skopeo sync` log.

mtrmac commented on copy/copy.go:
```diff
@@ -361,6 +365,44 @@ func supportsMultipleImages(dest types.ImageDestination) bool {
+	}
+
+	// Destination and source manifests, types and digests should all be equivalent
+	return true, destManifest, destManifestType, srcManifestDigest, nil
```
(Non-blocking: Using `destManifestDigest` here is exactly equivalent, but maybe a tiny bit easier to follow.)

mtrmac commented on copy/copy.go:
```diff
@@ -646,6 +688,27 @@ func (c *copier) copyOneImage(ctx context.Context, policyContext *signature.Poli
 	ic.diffIDsAreNeeded = ic.diffIDsAreNeeded || (isEncrypted(src) && ic.c.ociDecryptConfig != nil) || ic.c.ociEncryptConfig != nil
 
+	options.OptimizeDestinationImageAlreadyExists = true // TODO: PR: Remove forced enablement for PR testing
+	// If enabled, fetch and compare the destination's manifest. And as an optimization skip updating the destination iff equal
```
This does not make much sense just above the `isSigned` computation.

It might make sense to document the purpose of the whole section of the code, OTOH maybe just the `if options.Optimize…` (if it becomes top-level, as suggested below) with no comment is enough for that.

mtrmac commented on copy/copy.go:
```diff
@@ -646,6 +688,27 @@ func (c *copier) copyOneImage(ctx context.Context, policyContext *signature.Poli
+	options.OptimizeDestinationImageAlreadyExists = true // TODO: PR: Remove forced enablement for PR testing
+	// If enabled, fetch and compare the destination's manifest. And as an optimization skip updating the destination iff equal
+	// To be enabled, options.OptimizeDestinationImageAlreadyExists must be true, the image must not be signed or encryted. And must
+	// not have any pending manifest updates
```
(I’m not sure that this comment helps much — it just repeats the code, doesn’t explain anything.)

mtrmac commented on copy/copy.go:
```diff
@@ -646,6 +688,27 @@ func (c *copier) copyOneImage(ctx context.Context, policyContext *signature.Poli
+	// not have any pending manifest updates
+	isSigned := len(sigs) > 0 || options.SignBy != "" // TODO: Consider allowing signatures updates only and skipping the image's layers/manifest copy if possible
+
+	logrus.Debugf("Evaluating image manifest check optimzation: options.OptimizeDestinationImageAlreadyExists=%t !isSigned=%t !destRequiresOciEncryption=%t ic.noPendingManifestUpdates()=%t", options.OptimizeDestinationImageAlreadyExists, !isSigned, !destRequiresOciEncryption, ic.noPendingManifestUpdates())
```
Please wrap all of this in `if options.OptimizeDestinationImageAlreadyExists`, so that this log line does not appear if the caller does not opt in (and the presence of that log line is enough to indicate the value of that flag, so it doesn’t need to be logged separately); that way we don’t need to compute `isSigned` if the caller does not opt in.

mtrmac replied:
That debug line could be polished a bit — e.g. probably log `isSigned` instead of `!isSigned`, and the like.

mtrmac commented on copy/copy.go:
```diff
@@ -646,6 +688,27 @@ func (c *copier) copyOneImage(ctx context.Context, policyContext *signature.Poli
+	// If enabled, fetch and compare the destination's manifest. And as an optimization skip updating the destination iff equal
+	// To be enabled, options.OptimizeDestinationImageAlreadyExists must be true, the image must not be signed or encryted. And must
+	// not have any pending manifest updates
+	isSigned := len(sigs) > 0 || options.SignBy != "" // TODO: Consider allowing signatures updates only and skipping the image's layers/manifest copy if possible
```
This is not quite `isSigned`, rather  `shouldUpdateSigs` or something like that.

mtrmac reviewed (commented):
I’m sorry about the terribly late review.

The `targetInstance` point is the only major outstanding implementation concern; the rest is mostly stylistic.

gjenkins8 commented:
Thanks! @mtrmac, thanks for the comprehensive review!

I've fixed the `targetInstance` issue (but some testing thoughts below), and cleaned up the minor style/logging/misc issues you flagged.

The only (very minor) comment/thought I had was: https://github.com/containers/image/pull/1041#discussion_r570042508

And there is the omnipresent: https://github.com/containers/image/pull/1041/files#r500609569

**Additional Test Coverage**:

Both these changes https://github.com/containers/image/pull/1041/commits/702784dc9f95516c7e6da648f01e91f4828c508f & https://github.com/containers/image/pull/1041/commits/a2287d58ff33d2aeef9f9afc210e9131bfd74455 were to fix implementation issues/bugs.

The first, addressed an issue where an image will failed to be "retagged" to another already existing image. e.g. if you have `latest -> 1.23.0`, and you update in the source repository `latest -> 1.34.0`, this "update" would be skipped when syncing to the destination.

The second, IIUC, is to ensure the correct manifest for a manifest list is addressed.

Since the existing `skopeo` test suite passes this PR without these changes, it seems a good idea to add coverage of these scenarios? To prevent future regression. I can attempt this, as a PR to `skopeo` repo. If you think that is a good idea?

mtrmac commented on copy/copy.go:
```diff
@@ -646,6 +689,26 @@ func (c *copier) copyOneImage(ctx context.Context, policyContext *signature.Poli
+	if options.OptimizeDestinationImageAlreadyExists {
+		shouldUpdateSigs := len(sigs) > 0 || options.SignBy != "" // TODO: Consider allowing signatures updates only and skipping the image's layers/manifest copy if possible
+
+		logrus.Debugf("Evaluating image manifest check optimzation: shouldUpdateSigs=%t destRequiresOciEncryption=%t ic.noPendingManifestUpdates()=%t", shouldUpdateSigs, destRequiresOciEncryption, ic.noPendingManifestUpdates())
```
```suggestion
		logrus.Debugf("Evaluating image manifest check optimization: shouldUpdateSigs=%t destRequiresOciEncryption=%t ic.noPendingManifestUpdates()=%t", shouldUpdateSigs, destRequiresOciEncryption, ic.noPendingManifestUpdates())
```
at least, or
```suggestion
		logrus.Debugf("Checking if we can avoid the copy: shouldUpdateSigs=%t destRequiresOciEncryption=%t ic.noPendingManifestUpdates()=%t", shouldUpdateSigs, destRequiresOciEncryption, ic.noPendingManifestUpdates())
```
, or even
```suggestion
		logrus.Debugf("Checking if we can avoid the copy: has signatures=%t, OCI encryption=%t, no manifest updates=%t", shouldUpdateSigs, destRequiresOciEncryption, ic.noPendingManifestUpdates())
```
I don’t have a strong preference.

gjenkins8 replied:
sure, I did the latter:
`		logrus.Debugf("Checking if we can avoid the copy: has signatures=%t, OCI encryption=%t, no manifest updates=%t", shouldUpdateSigs, destRequiresOciEncryption, noPendingManifestUpdates)`

mtrmac commented on copy/copy.go:
```diff
@@ -646,6 +689,26 @@ func (c *copier) copyOneImage(ctx context.Context, policyContext *signature.Poli
+		shouldUpdateSigs := len(sigs) > 0 || options.SignBy != "" // TODO: Consider allowing signatures updates only and skipping the image's layers/manifest copy if possible
+
+		logrus.Debugf("Evaluating image manifest check optimzation: shouldUpdateSigs=%t destRequiresOciEncryption=%t ic.noPendingManifestUpdates()=%t", shouldUpdateSigs, destRequiresOciEncryption, ic.noPendingManifestUpdates())
+		if !shouldUpdateSigs && !destRequiresOciEncryption && ic.noPendingManifestUpdates() {
```
`ic.noPendingManifestUpdates` is a bit expensive, it would be nice (and safer) to only compute it once.

gjenkins8 replied:
Good spot. Apologies, not enough thinking on my behalf. I've factored out.

mtrmac reviewed (commented):
@vrothberg Can I hand this off to you?

vrothberg commented:
> @vrothberg Can I hand this off to you?

Sure. I added it to my list, thanks!

mtrmac reviewed (commented):
LGTM (apart from https://github.com/containers/image/pull/1041/files#r500609569 ). Thanks again!

gjenkins8 commented:
> LGTM (apart from https://github.com/containers/image/pull/1041/files#r500609569 ). Thanks again!

@mtrmac / @vrothberg -- 1372f3c removes the enabling of `OptimizeDestinationImageAlreadyExists` I had included for testing.

gjenkins8 commented:
And oops, sorry. Thanks for the review @mtrmac !

vrothberg commented on copy/copy.go:
```diff
@@ -361,6 +365,45 @@ func supportsMultipleImages(dest types.ImageDestination) bool {
 }
 
+// compareImageDestinationManifestEqual compares the `src` and `dest` image manifests (reading the manifest from the
+// (possibly remote) destination). Returning true and the image's manifest, type and digest if they compare equal.
```
```suggestion
// (possibly remote) destination). Returning true and the destination's manifest, type and digest if they compare equal.
```

vrothberg commented on copy/copy.go:
```diff
@@ -361,6 +365,45 @@ func supportsMultipleImages(dest types.ImageDestination) bool {
+		return false, nil, "", "", errors.Wrapf(err, "Error calculating manifest digest")
+	}
+
+	logrus.Debugf("Source and destination manifest digests: %v vs. %v", srcManifestDigest, destManifestDigest)
```
```suggestion
	logrus.Debugf("Comparing source and destination manifest digests: %v vs. %v", srcManifestDigest, destManifestDigest)
```
Would be a bit easier for me when skimming logs.

vrothberg commented on copy/copy.go:
```diff
@@ -646,6 +689,26 @@ func (c *copier) copyOneImage(ctx context.Context, policyContext *signature.Poli
+			}
+
+			if isSrcDestManifestEqual {
+				c.Printf("Copy skipped, expected image already present\n")
```
```suggestion
				c.Printf("Skipping: image already present at destination\n")
```
"Skipping" for consistency with other logs on writer.

vrothberg commented on copy/copy.go:
```diff
@@ -646,6 +689,26 @@ func (c *copier) copyOneImage(ctx context.Context, policyContext *signature.Poli
+		shouldUpdateSigs := len(sigs) > 0 || options.SignBy != "" // TODO: Consider allowing signatures updates only and skipping the image's layers/manifest copy if possible
+		noPendingManifestUpdates := ic.noPendingManifestUpdates()
+
+		logrus.Debugf("Checking if we can avoid the copy: has signatures=%t, OCI encryption=%t, no manifest updates=%t", shouldUpdateSigs, destRequiresOciEncryption, noPendingManifestUpdates)
```
```suggestion
		logrus.Debugf("Checking if we can skip copying: has signatures=%t, OCI encryption=%t, no manifest updates=%t", shouldUpdateSigs, destRequiresOciEncryption, noPendingManifestUpdates)
```

vrothberg reviewed (commented):
Thanks for working on this, @gjenkins8!

I only have minor nits and the request to squash the commits into a single one.

vrothberg commented on copy/copy.go:
```diff
@@ -646,6 +689,26 @@ func (c *copier) copyOneImage(ctx context.Context, policyContext *signature.Poli
+		noPendingManifestUpdates := ic.noPendingManifestUpdates()
+
+		logrus.Debugf("Checking if we can avoid the copy: has signatures=%t, OCI encryption=%t, no manifest updates=%t", shouldUpdateSigs, destRequiresOciEncryption, noPendingManifestUpdates)
+		if !shouldUpdateSigs && !destRequiresOciEncryption && noPendingManifestUpdates {
```
Can you add a `// TODO: Consider compression changes` here?

Just in case we want further optimizations in the future.

mtrmac replied:
IIRC (a drive-by comment without reading much else), requested compression changes are currently ignored anyway if `TryReusingBlob` finds a match — and if we find the full manifest, that should always be the case. So requested compression changes should not be a reason to skip this optimization.

gjenkins8 replied:
Thanks for the explanation @mtrmac

To paraphrase/If I understand correctly: further work would be needed for change in compression to be effected in a copy? Only then, would it make sense to disable the optimization (or more specifically, then it would be required to disable the optimization).

I've added the `// TODO: ..` as per @vrothberg's suggestion. But I don't see anything else required here(?)

mtrmac replied:
No, I’m saying that if the destination manifest exists, any requested re-compression via `types.SystemContext.Compression{Operation,Algorithm}` wouldn’t happen _anyway_, so there is nothing to be done here.

OTOH I’m not 100% sure what “consider compression changes” refers to, @vrothberg can you clarify it into something more precise / actionable?

vrothberg replied:
I am cool to drop the comment if it's causing more confusion than good.  My intention was to leave some breadcrumbs in case `skopeo sync` would support compression changes.  As far as I understand the code, compression changes would be ignored and @mtrmac's comment above confirm.

gjenkins8 replied:
I've remove the comment. Thanks

gjenkins8 commented:
@vrothberg -- thanks for the review. I've updated with your suggested changes. And squashed the commits.

vrothberg commented:
> @vrothberg -- thanks for the review. I've updated with your suggested changes. And squashed the commits.
> 
> (and note: [CI is failing](https://travis-ci.org/github/containers/image/builds/758654465?utm_source=github_status&utm_medium=notification) due to #1107 -- I'll re-run shortly)

Thanks for your work!  When repushing, could you also rebase?

gjenkins8 commented:
> > @vrothberg -- thanks for the review. I've updated with your suggested changes. And squashed the commits.
> > (and note: [CI is failing](https://travis-ci.org/github/containers/image/builds/758654465?utm_source=github_status&utm_medium=notification) due to #1107 -- I'll re-run shortly)
> 
> Thanks for your work! When repushing, could you also rebase?

@vrothberg -- done

vrothberg reviewed (commented):
LGTM, thanks!
